features/370-auth-ui/web-ui.md

```diff
@@ -0,0 +1,179 @@
+
+```
+|---------------------------|
+| Login with Google         |
+|---------------------------|
+| Login with Facebook       |
+|---------------------------|
+
+              Or
+
+|--------------------------------------|  |----------|
+| Enter in your email or username here |  | Continue |
+|--------------------------------------|  |----------|
+
+Login with a phone number instead.
+```
+
+If the user choose to sign in with an IdP, the user is redirected to the authorization endpoint of that IdP. If the user continues with a login ID, the user is directed to the password page.
+
+Separating the login ID and the password into two pages allow future improvements, such as [Open ID Provider Issuer Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery). For example, the user can type in their Google account name and then be redirected to the authorization endpoint of Google.
+
+### The sign in password page
+
+The sign in password page displays a visibility toggleable password field *WITHOUT* password requirements.
+
+```
+|------------------------|  |----------|
+| Create a password here |  | Continue |
+|------------------------|  |----------|
+```
+
+### The sign up page
+
+The sign up page displays the first login ID key default. Other login ID keys are available to choose.
+
+```
+Sign up with email
+|--------------------------|  |----------|
+| Enter in your email here |  | Continue |
+|--------------------------|  |----------|
+
+Sign up with phone instead.
+Sign up with username instead.
+```
+
+### The sign up password page
+
+The sign up password page displays a visibility toggleable password field with password requirements.
+
+```
+|------------------------|  |----------|
+| Create a password here |  | Continue |
+|------------------------|  |----------|
+
+- [ ] At least one digit
+- [ ] At least one uppercase English character
+- [ ] At least one lowercase English character
+- [ ] At least one symbols ~`!@#$%^&*()-_=+[{]}\|;:'",<.>/?
+- [ ] At least 8 characters long
+```
+
+### The settings endpoint
+
+The settings endpoint `/settings` allows the user to manage their account. Features include but not limited to:
+
+- Manage sessions at Skygear Auth
+- Manage OAuth authorizations
+- Manage OAuth grants
+- Update metadata
+- Change password
+- Configure multi-factor authentication
+- Manage identities (e.g. login IDs and connections with external IdPs)
+
+### The verification endpoint
+
+The verification endpoint `/verify` allows the user to verify their login ID. This page requires authenticated user. Unauthenticated access is redirected to the root endpoint with `redirect_uri=/verify`. Verification is not a strict requirement. For example, if auto send verification email is enabled, the email is sent behind the scene and the user is redirected to the application immediately after performing the Authorization Code Flow successfully.
```

Comment:
> Carmen: Would like to know how user do verification for native app? User will complete the authorization flow without the app, and user will need to login again outside the app in browser?

# Option 1: Make the authorization endpoint enforce verification with an option

1. To let the developer know whether the user is considered verified or not, we should introduce a new claim `is_verified` to UserInfo.
2. In addition to [accepting Access Token in the UserInfo endpoint](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo), the endpoint also accepts IdP session cookie.
3. Add SDK method to call the UserInfo endpoint. If Access Token is present, use it as bearer token, otherwise assume IdP session cookie is present.
4. The developer can then call the UserInfo endpoint via SDK to obtain a UserInfo Response, which contains `is_verified`.
5. If `is_verified` is false, the developer call `/oauth2/authorize?verify=true`, which enforce verification

# Option 2: Verification is mandatory

The only difference from Option 1 is that `verify=true` in Step 5 is unnecessary.

# Whether addLoginID and login ID verification is atomic

In the Auth API, the two are separate operations. An unverified login ID can be used in the same way as verified login ID without any restriction. Should we make adding login ID more similar to adding an authenticator? That is, the newly added login ID must be "verified" before it is usable. But only `email` and `phone` login ID type can be verified. `username` and `raw` can never be verified.

---

Thoughts? @carmenlau @kiootic

Reply:
I think in native app, after authentication, there would be a pair of access/refresh token in the app, and a session token in cookies of external browser. In this case, developer can simply call userinfo endpoint to check whether user is verified, then open verify endpoint in external browser directly if not verified.

Reply:
> I think in native app, after authentication, there would be a pair of access/refresh token in the app, and a session token in cookies of external browser.

Step 2 is for web application, where no access token is obtained after authentication. But the developer still need to initiate a CORS request to the UserInfo endpoint.

> In this case, developer can simply call userinfo endpoint to check whether user is verified, then open verify endpoint in external browser directly if not verified.

So the authorization endpoint does not handle verification at all?

Reply:
I think if we want to integrate user verification into authentication flow, we should require it automatically depending on server-side configuration, so both option 1 and 2 is not good enough. I think we don't need to support it at the moment?

Reply:
Option 2 integrates user verification into the authentication flow so it is solely by server-side configuration. The problem is, the user can open the settings endpoint and add a unverified login ID. This could make the user become unverified. So even if we have option 2, the developer still have to check UserInfo in suitable places to ensure the user is verified.

Reply:
I think the role of OP is to assert the identity of user to RP. To avoid complicated logic at RP side and mistaken interpretation of claims, we should have a config flag for user verification (name for reference only):
- always_do_verification: always verify user in authentication flow, `*_verified` claims are always true.
- no_verify: don't verify user in authentication flow, disable verify APIs, `*_verified` claims are always false.
- use_api_to_verify: don't verify user in authentication flow, enable verify APIs, `*_verified` claims can be true/false.

I think we don't need to support it at the moment?

Reply:
I didn't read the discussion in details, only saw this at Slack... I think we can just let users use (admin?) API to verify first? And incorporate verification into auth flow later?

Reply:
@kiootic The reason I proposed option 1 is because the verification endpoint is like the authorization endpoint plus verification. It requires `redirect_uri` so it can redirect back to the application.

Reply:
I think option 1 would mix up the concerns of OAuth/OIDC and verification. The `authorize` endpoint is for obtaining consent from end-user to access protected resource (i.e. APIs in OAuth and identity information in OIDC); performing authentication is just a side-effect. It would be better to have a proprietary endpoint for presenting the verification UI if it does not fit into the authentication flow.

Reply:
See if https://github.com/SkygearIO/features/pull/425/commits/b2f509b807dd7cf5e9deb53bd805f21d2344fec0 captures the conclusion here.